### PR TITLE
[mask rom] Add mask_rom_test.h for mask ROM unit tests.

### PR DESCRIPTION
**Pull request by alphan**
This change introduces `mask_rom_test.h`, which provides classes for mocking dependencies of and unit testing mask ROM modules. Please see [this document](https://docs.google.com/document/d/1D_3F2FoeCW2GrW_nDIg0mFj8vZ6k2OHHyzX8t2fJvHc/edit?usp=sharing) for more information.

Signed-off-by: Alphan Ulusoy <alphan@google.com>

### CODE REVIEW

File: sw/device/lib/testing/mask_rom_test.h
Comment (mcy) [inline sw/device/lib/testing/mask_rom_test.h:31]: I'm confused... why is this a `thread_local static`, rather than a non-static member variable? There's no guarantee that tests will be executed on distinct threads, so this won't provide any sort of isolation.

Based on the code you've written, I think it makes sense to design mask ROM APIs that we want to be able to mock to pass around some kind of "context" void*, which we can use to inject mocks in tests.

An alternative that might more clearly make this API generic might be something like this:
```cc
template <typename T>
class InjectContext {
 public:
  template<typename Args...>
  InjectContext(T *&global_ptr, Ts&&... args) : place_(place) {
    // Optional: CHECK(global_ptr == nullptr)?
    // You could ostensibly save the old value of global_ptr and swap it
    // back in the dtor.
    place_ = new T(std::forward(args...));
  }
  ~InjectContext() { delete place_; place_ = nullptr; }

  T &operator*() {
    return *place_;
  }
  T *operator->() {
    return place_;
  }

 private:
   T *&place_;
}
```

Then, in your test, you could write
```cc
thread_local MyMock* my_mock;
/// ...
InjectContext ctx(&my_mock, MyMock());
EXPECT_CALL(*ctx, ...);
```

Thus, we can see in each test much more clearly where each mock is injected. You could even imagine wrapping the `thread_local T*` in another type, e.g. `GlobalContext` or something, and give it a member function like `->Inject()` that constructs a new `InjectContext`:
```cc
GlobalContext<MyMock> my_mock;
/// ...
auto ctx = my_mock.Inject(MyMock());
```

What do you think?

**Comment by lenary** [comment]
My feeling is this is a reasonable approach, so long as it forms part of a larger testing approach which has to include on-device tests, compiled for ibex and OT specifically.

- I think this approach is good for testing the functionality of the main mask_rom boot function, which is its intention.
- I agree with the sentiment that testing the main mask rom boot function using only the mock_mmio functionality is going to ensure that developers fail to look closely at how tests change, and details will get lost in the other noise. 
- I think we will test the second-level functions (called by the main mask_rom function) with the same approach, either by using mock_mmio, or by intercepting the calls to DIFs themselves.

File: sw/device/lib/testing/mask_rom_test.h
Comment (alphan) [reply sw/device/lib/testing/mask_rom_test.h:31]: > I'm confused... why is this a thread_local static, rather than a non-static member variable? There's no guarantee that tests will be executed on distinct threads, so this won't provide any sort of isolation.

This is a `static` because mask rom functions don't have an argument like the DIF functions do that we can use to inject a mock. This is basically what the googletest doc recommends for [sharing resources between tests](https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#sharing-resources-between-tests-in-the-same-test-suite). Regarding `thread_local`, you are right, it is not really needed since AFAIK tests don't run in parallel, this is left over from earlier trials.

> Based on the code you've written, I think it makes sense to design mask ROM APIs that we want to be able to mock to pass around some kind of "context" void*, which we can use to inject mocks in tests.

I would ideally prefer this too but my understanding is that we want to avoid any sort of test-related code in mask rom functions. Apologies for not clearly stating this in the PR. If having this `context` argument and passing it around to every function is acceptable, then we obviously don't need the static member. @gdk, @lenary, any thoughts?

> An alternative that might more clearly make this API generic might be something like this

If we have an argument in all mask rom functions which we can use to inject the mock object in tests, why would we need this? IMHO, just calling the mask rom functions with the mock and using the mock in their definitions in the test file to forward the calls to its methods would be clearer. Is there something that I'm missing here?

Comment (lenary) [reply sw/device/lib/testing/mask_rom_test.h]: I'm not very happy with a `void* context` object appearing everywhere in the mask rom, precisely as @mcy points out: it's there for test-only reasons, and we've said we don't want to do that. I think the current approach, as sketched will work for us without needing this.

**Comment by alphan** [comment]
Updated the PR.

@mcy, I made a small change to `GlobalMock<T>` to make it a subclass like `testing::StrictMock` instead of a base. PTAL.

**Review by silvestrst — COMMENTED** [review]
@alphan, I think this looks great, thank you!

There are couple of things that I think we need to give some thought:

- This works well, as long as the Mask ROM modules don't ever interact (which I think is the case, apart from the main Mask ROM module.

- Are we not unittesting the main Mask ROM module?

Overall looks great, and additional thanks for a comprehensive document and code examples.

**Comment by silvestrst** [comment]
> I think this approach is good for testing the functionality of the main mask_rom boot function, which is its intention.

I am not sure this is correct, I think the purpose of this change is to add the framework for testing all the Mask ROM modules, but as it stands - it doesn't cover main boot function (as the main boot module will call into the other "sub-modules" which are not mocked by this change, and not intended to be mocked as I understand)?

File: sw/device/tests/mask_rom/mask_rom_fake_keys_sig_unittest.cc
Comment (alphan) [reply sw/device/tests/mask_rom/mask_rom_fake_keys_sig_unittest.cc:5]: Thanks!

File: sw/device/tests/mask_rom/mocks/dif_fake_hmac_mock.h
Comment (alphan) [reply sw/device/tests/mask_rom/mocks/dif_fake_hmac_mock.h:9]: I agree with both of your points but looking at the dependency graph of `boot()`, I think that automating this busywork is still worth it. I hope to give this a try after the keys & signature module lands and see how everyone feels about it.

**Comment by alphan** [comment]
> it doesn't cover main boot function (as the main boot module will call into the other "sub-modules" which are not mocked by this change, and not intended to be mocked as I understand)?

@silvestrst, while the example in this PR doesn't have a mock mask ROM module, this approach covers all mask ROM modules, including the `boot()` function (from the doc):

> The testing strategy for an individual mask ROM module is to ensure that for a set of inputs, its public functions
> - Make the expected calls to its direct dependencies with the expected arguments, and
> - Produce the expected results in the form of return and out parameter values.
>
>Consequently, the testing strategy for the entire mask ROM is to perform this for all modules until the entire mask ROM is covered.

Please note that the "dependencies" mentioned here can be DIF libraries or other mask ROM modules.

**Comment by silvestrst** [comment]
As per chat with @alphan , I have misread the document slightly - we are testing the boot function in the same way we are testing the individual modules. To test the boot function (main mask_rom module) - we would need to mock all the modules it uses.

**Review by silvestrst — APPROVED** [review]

**Review by lenary — APPROVED** [review]
LGTM. I know the code has worked out as more complex than we wanted, but I think that's just the cost of doing mocking in C. We'll survive.

File: sw/device/lib/testing/mask_rom_test.h
Comment (mcy) [inline sw/device/lib/testing/mask_rom_test.h:55]: Why are these marked as protected? Just leave them as public, so that the error message makes a bit more sense (i.e., we want the error to be "can't copy", not "copy ctor not visible".

File: sw/device/lib/testing/mask_rom_test.h
Comment (mcy) [inline sw/device/lib/testing/mask_rom_test.h:18]: This is just a language-lawyerey nit, but C++ doesn't use terminology like "subclass" or "instance". Personally, I would have written "derived class" and "value", but it's up to you.

File: sw/device/lib/testing/mask_rom_test.h
Comment (mcy) [inline sw/device/lib/testing/mask_rom_test.h:11]: I feel like GlobalMock is a sufficiently generic type (i.e., I want to use it for CSR mocking) that it probably doesn't want to be closely associated with the Mask ROM name.

File: sw/device/lib/testing/mask_rom_test.h
Comment (mcy) [inline sw/device/lib/testing/mask_rom_test.h:37]: Honestly, I feel like CTRP is nicer here, but I understand why you'd do this, to be consistent with some of the googletest "mixins".

File: sw/device/lib/testing/mask_rom_test.h
Comment (mcy) [inline sw/device/lib/testing/mask_rom_test.h:90]: I'm still not sure if I'm happy with this default. It feels like testing for function calls doesn't need to be by-default in sequence like MMIO calls are, and I feel it would make the tests more consistent with usual googletest style to be out-of-order by default.

I'm mostly trying to stick to googletest convention, but I understand if we can't in some places because of our slightly weird use of it.

(Separately, I think that the `Unordered<MaskRomTest>` idiom is slightly bizarre, since `Unordered<T>` is meaningless except for `T = MaskRomTest`; at that point, you might as well not use MaskRomTest at all...)

File: sw/device/tests/mask_rom/mask_rom_fake_keys_sig_unittest.cc
Comment (mcy) [inline sw/device/tests/mask_rom/mask_rom_fake_keys_sig_unittest.cc:15]: Do not use using directives (i.e., `using namespace`). This breaks name resolution in weird ways, and is banned in Google C++.

See https://google.github.io/styleguide/cppguide.html#Namespaces.

(I know this is example code, but I feel it's very important to point out that using namespace is a gigantic footgun. https://abseil.io/tips/153 has some details.)

File: sw/device/tests/mask_rom/mocks/dif_fake_hmac_mock.h
Comment (mcy) [reply sw/device/tests/mask_rom/mocks/dif_fake_hmac_mock.h:9]: I suspect this should be really easy, given our hyper-strict interface rules. We just need to parse function protypes and nothing else. Definitely worth a shot.

File: sw/device/lib/testing/mask_rom_test.h
Comment (mcy) [inline sw/device/lib/testing/mask_rom_test.h:46]: Oh, I just realized something super obnoxious about this setup: if someone accidentally converts a GlobalMock<T>* into a T* and then deletes it, this dtor won't run if T* doesn't have a virtual destructor. I *highly* doubt this will be an issue in practice, but perhaps consider including the following as a sanity check...?

```
static_assert(std::has_virtual_destructor<T>::value, "Mock class must have a virtual destructor.");
```

File: sw/device/lib/testing/mask_rom_test.h
Comment (mcy) [inline sw/device/lib/testing/mask_rom_test.h:39]: Optional: make this a forwarding ctor, which I think is good practice for wrappers like this.

```cpp
template <typename... Args>
GlobalMock(Args&&... args) : T(std::forward(args)...) { ... }
```

File: sw/device/lib/testing/mask_rom_test.h
Comment (alphan) [reply sw/device/lib/testing/mask_rom_test.h:11]: I agree, we can move it out when we use it somewhere else.

File: sw/device/lib/testing/mask_rom_test.h
Comment (alphan) [reply sw/device/lib/testing/mask_rom_test.h:39]: Didn't do this initially because most likely we won't have a mock that needs this, esp. if we use a script to generate them. Done.

File: sw/device/lib/testing/mask_rom_test.h
Comment (alphan) [reply sw/device/lib/testing/mask_rom_test.h:46]: Similar to above, but this is nice to have in case we use this in other contexts. Done.

File: sw/device/lib/testing/mask_rom_test.h
Comment (alphan) [reply sw/device/lib/testing/mask_rom_test.h:90]: I understand, but we can always revisit this and I think this is a reasonable place to start. Let's keep an eye on the tests and see which case is used most. Regarding the mixin, yes, it feels pointless since `MaskRomTest` is such a thin class but I suggest we move forward and see if need it at all.

**Comment by alphan** [comment]
Thanks for your feedback @silvestrst, @lenary, and @mcy! I'll merge this once CI is done.

**Pull request merged**
